stonesoup/predictor/asd.py

```diff
@@ -0,0 +1,275 @@
+                state_pred_m, predict_over_interval=time_interval_m_to_m_plus_1, **kwargs)
+
+            # transitions for timestamp m
+            transition_matrix_m = self._transition_matrix(
+                prior=prior_at_t, time_interval=predict_over_interval, **kwargs)
+            # transitions for timestamp m+1
+            transition_matrix_m_plus_1 = self._transition_matrix(
+                prior=state_pred_m, time_interval=time_interval_m_to_m_plus_1,
+                **kwargs)
+
+            t_minus2t = slice((t_index-1) * ndim, t_index * ndim)
+            x_m_plus_1_given_k = prior.multi_state_vector[t_minus2t]
+            x_diff = x_m_plus_1_given_k - x_pred_m_plus_1
+
+            p_m_plus_1_given_k = prior.multi_covar[t_minus2t, t_minus2t]
+            p_diff = p_m_plus_1_given_k - p_pred_m_plus_1
+
+            W = p_pred_m @ transition_matrix_m_plus_1.T @ np.linalg.inv(p_pred_m_plus_1)
+            x_pred_m = x_pred_m + W@x_diff
+            p_pred_m_cr = p_pred_m + W@p_diff@W.T
+
+            # build full state
+            x_pred = np.concatenate([prior.multi_state_vector[:t_index * ndim],
+                                     x_pred_m,
+                                     prior.multi_state_vector[t_index * ndim:]])
+
+            P_right_lower = prior.multi_covar[t_index * ndim:, t_index * ndim:]
+
+            # add new correlation matrix with the present time step
+            pred_from_corr_matrices = correlation_matrices[timestamp_from_which_is_predicted]
+            pred_from_corr_matrices['P_pred'] = p_pred_m
+            pred_from_corr_matrices['F'] = transition_matrix_m
+            pred_from_corr_matrices['PFP'] = (
+                    pred_from_corr_matrices['P'] @ transition_matrix_m.T @ np.linalg.inv(p_pred_m))
+
+            correlation_matrices[timestamp] = {}
+            correlation_matrices[timestamp]['F'] = transition_matrix_m_plus_1
+            correlation_matrices[timestamp]['P_pred'] = p_pred_m_plus_1
+            correlation_matrices[timestamp]['P'] = p_pred_m_cr
+            correlation_matrices[timestamp]['PFP'] = \
+                p_pred_m_cr @ transition_matrix_m_plus_1.T @ np.linalg.inv(p_pred_m_plus_1)
+
+            # resort the dict
+            correlation_matrices = OrderedDict(sorted(correlation_matrices.items(), reverse=True))
+
+            # generate prediction matrix
+            p_pred = np.zeros((ndim * (prior.nstep+1), ndim * (prior.nstep+1)))
+            p_pred[(t_index+1) * ndim:, (t_index+1) * ndim:] = P_right_lower
+
+            # get all timestamps which has to be recalculated beginning
+            # with the newest one
+            timestamps = sorted(prior.timestamps + [timestamp], reverse=True)
+            timestamps_to_recalculate = [ts for ts in timestamps if ts >= timestamp]
+            covars = \
+                [prior.multi_covar[i * ndim:(i+1) * ndim, i * ndim:(i+1) * ndim]
+                 for i in range(t_index)]
+            covars.append(p_pred_m)
```

Comment:
I think this should be `p_pred_m_cr` rather than `p_pred_m`. Changing this, it seems to give better and more stable results, whereas without changing it, the estimates looking wrong.

Reply:
I've changed this in ecfd1a6, which also includes changing the covariances stored in correlation matrices, as I believe this were the wrong way around.

Reply:
Sorry for the late answer. You are right. This was an error in my implementation.